README.md
# bytes_data_builder

Builds a list of bytes, allowing bytes, fixed-width integers, floats and lists of bytes to be added at the end.

## Usage

A simple usage example:

    import 'package:bytes_data_builder/bytes_data_builder.dart';

    main() {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt8(0);
      List<int> result = builder.takeBytes();
    }


CHANGELOG.md
# Changelog

## 0.1.1

- Initial version, created by Kurt Morris

## 0.1.2

- Mistake corrected in README.md

## 0.1.3

- CHANGELOG.md replaced with proper change log.

## 0.1.4

- Added GitHub Repository, made description more aesthetically pleasing.

test/BytesDataBuilder_test.dart
// Copyright (c) 2016, Kurt Morris. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:math';
import 'dart:typed_data';
import 'package:collection/collection.dart';
import 'package:bytes_data_builder/bytes_data_builder.dart';
import 'package:test/test.dart';

void main() {
  group('Proper byte logic (Big Endian, Little Endian, Signed, Unsigned)', () {
    Function eq;
    int int8data;
    int int16data;
    int int32data;
    int int64data;
    int Uint8data;
    int Uint16data;
    int Uint32data;
    int Uint64data;

    setUp(() {
      eq = const ListEquality().equals;
      Random rand = new Random();
      int8data = rand.nextInt(255) - 127;
      int16data = rand.nextInt(65535) - 32767;
      int32data = rand.nextInt(4294967295) - 2147483647;
      int rand64 = rand.nextInt(4294967295) * rand.nextInt(4294967295);
      int64data = rand64 - 9223372036854775807;
      Uint8data = rand.nextInt(255);
      Uint16data = rand.nextInt(65535);
      Uint32data = rand.nextInt(4294967295);
      Uint64data = rand.nextInt(4294967295) * rand.nextInt(4294967295);
    });

    test('int8', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt8(int8data);
      ByteData bdata = new ByteData(1);
      bdata.setInt8(0, int8data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int16', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt16(int16data);
      ByteData bdata = new ByteData(2);
      bdata.setInt16(0, int16data);
      Uint8List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int32', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt32(int32data);
      ByteData bdata = new ByteData(4);
      bdata.setInt32(0, int32data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int64', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt64(int64data);
      ByteData bdata = new ByteData(8);
      bdata.setInt64(0, int64data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint8', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint8(Uint8data);
      ByteData bdata = new ByteData(1);
      bdata.setUint8(0, Uint8data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint16', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint16(Uint16data);
      ByteData bdata = new ByteData(2);
      bdata.setUint16(0, Uint16data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint32', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint32(Uint32data);
      ByteData bdata = new ByteData(4);
      bdata.setUint32(0, Uint32data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint64', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint64(Uint64data);
      ByteData bdata = new ByteData(8);
      bdata.setUint64(0, Uint64data);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int16LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt16(int16data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(2);
      bdata.setInt16(0, int16data, Endianness.LITTLE_ENDIAN);
      Uint8List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int32LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt32(int32data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(4);
      bdata.setInt32(0, int32data, Endianness.LITTLE_ENDIAN);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('int64LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addInt64(int64data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(8);
      bdata.setInt64(0, int64data, Endianness.LITTLE_ENDIAN);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint16LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint16(Uint16data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(2);
      bdata.setUint16(0, Uint16data, Endianness.LITTLE_ENDIAN);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint32LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint32(Uint32data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(4);
      bdata.setUint32(0, Uint32data, Endianness.LITTLE_ENDIAN);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
    test('Uint64LE', () {
      BytesDataBuilder builder = new BytesDataBuilder();
      builder.addUint64(Uint64data, Endianness.LITTLE_ENDIAN);
      ByteData bdata = new ByteData(8);
      bdata.setUint64(0, Uint64data, Endianness.LITTLE_ENDIAN);
      List builderList = builder.takeBytes();
      List bdataList = bdata.buffer.asUint8List();
      expect(eq(builderList, bdataList), isTrue);
    });
  });
}


lib/bytes_data_builder.dart
// Copyright (c) 2016, Kurt Morris. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library bytes_data_builder;

import 'dart:typed_data';
import 'dart:io';

/// Builds a list of bytes, allowing bytes, integers, floats and
/// lists of bytes to be added at the end.
///
/// Used to efficiently collect bytes and lists of bytes.
class BytesDataBuilder implements BytesBuilder {
  final BytesBuilder _delegate;

  /// Construct a new empty [BytesDataBuilder].
  ///
  /// If [copy] is true, the data is always copied when added to the list. If
  /// it [copy] is false, the data is only copied if needed. That means that if
  /// the lists are changed after added to the [BytesDataBuilder], it may effect the
  /// output. Default is `true`.
  BytesDataBuilder({bool copy: true})
      : _delegate = new BytesBuilder(copy: copy);

  /// Appends [bytes] to the current contents of the builder.
  ///
  /// Each value of [bytes] will be bit-representation truncated to the range
  /// 0 .. 255.
  void add(List<int> bytes) => _delegate.add(bytes);

  /// Append [byte] to the current contents of the builder.
  ///
  /// The [byte] will be bit-representation truncated to the range 0 .. 255.
  void addByte(int byte) => _delegate.addByte(byte);

  /// Returns the contents of `this` and clears `this`.
  ///
  /// The list returned is a view of the internal buffer, limited to the
  /// [length].
  List<int> takeBytes() => _delegate.takeBytes();

  /// Returns a copy of the current contents of the builder.
  ///
  /// Leaves the contents of the builder intact.
  List<int> toBytes() => _delegate.toBytes();

  /// The number of bytes in the builder.
  int get length => _delegate.length;

  /// Returns `true` if the buffer is empty.
  bool get isEmpty => _delegate.isEmpty;

  /// Returns `true` if the buffer is not empty.
  bool get isNotEmpty => _delegate.isNotEmpty;

  /// Clear the contents of the builder.
  void clear() => _delegate.clear();

  /// Quick and slick way of converting multiple integer types to bytes.
  ///
  /// The [data] will be converted to bytes according to the amount
  /// of [bits], whether the bytes are [signed] bytes or not and of course
  /// the appropriate [Endianness] for the bytes.
  void _add(int data, int bits, bool signed, Endianness endian) {
    if (endian == Endianness.BIG_ENDIAN) {
      if (bits > 32) {
        addByte(signed ? data >> 56 : (data >> 56) & 0xff);
        addByte(signed ? data >> 48 : (data >> 48) & 0xff);
        addByte(signed ? data >> 40 : (data >> 40) & 0xff);
        addByte(signed ? data >> 32 : (data >> 32) & 0xff);
      }
      if (bits > 16) {
        addByte(signed ? data >> 24 : (data >> 24) & 0xff);
        addByte(signed ? data >> 16 : (data >> 16) & 0xff);
      }
      if (bits > 8) {
        addByte(signed ? data >> 8 : (data >> 8) & 0xff);
      }
      addByte(signed ? data : data & 0xff);
    } else if (endian == Endianness.LITTLE_ENDIAN) {
      addByte(signed ? data : data & 0xff);
      if (bits > 8) {
        addByte(signed ? data >> 8 : (data >> 8) & 0xff);
      }
      if (bits > 16) {
        addByte(signed ? data >> 16 : (data >> 16) & 0xff);
        addByte(signed ? data >> 24 : (data >> 24) & 0xff);
      }
      if (bits > 32) {
        addByte(signed ? data >> 32 : (data >> 32) & 0xff);
        addByte(signed ? data >> 40 : (data >> 40) & 0xff);
        addByte(signed ? data >> 48 : (data >> 48) & 0xff);
        addByte(signed ? data >> 56 : (data >> 56) & 0xff);
      }
    }
  }

  /// Appends four bytes to the current contents of the builder with
  /// the IEEE 754 single-precision binary floating-point (binary32)
  /// representation of the specified value.
  BytesDataBuilder addFloat32(double data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    ByteData bd = new ByteData(4);
    bd.setFloat32(0, data, endian);
    add(bd.buffer.asInt8List());
    return this;
  }

  /// Appends eight bytes to the current contents of the builder with
  /// the IEEE 754 single-precision binary floating-point (binary64)
  /// representation of the specified value.
  BytesDataBuilder addFloat64(double data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    ByteData bd = new ByteData(8);
    bd.setFloat64(0, data, endian);
    add(bd.buffer.asInt8List());
    return this;
  }

  /// Appends a byte to the current contents of the builder with
  /// the binary representation of the specified value,
  /// which must fit in a byte.
  BytesDataBuilder addInt8(int data) {
    addByte(data);
    return this;
  }

  /// Appends two bytes to the current contents of the builder with
  /// the two's complement binary representation of the specified value,
  /// which must fit in two bytes.
  BytesDataBuilder addInt16(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 16, true, endian);
    return this;
  }

  /// Appends four bytes to the current contents of the builder with
  /// the four's complement binary representation of the specified value,
  /// which must fit in four bytes.
  BytesDataBuilder addInt32(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 32, true, endian);
    return this;
  }

  /// Appends eight bytes to the current contents of the builder with
  /// the eight's complement binary representation of the specified value,
  /// which must fit in eights bytes.
  BytesDataBuilder addInt64(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 64, true, endian);
    return this;
  }

  /// Appends a byte to the current contents of the builder with
  /// the binary representation of the specified value,
  /// which must fit in a byte.
  BytesDataBuilder addUint8(int data) {
    addByte(data & 0xff);
    return this;
  }

  /// Appends two bytes to the current contents of the builder with
  /// the two's complement binary representation of the specified value,
  /// which must fit in two bytes.
  BytesDataBuilder addUint16(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 16, false, endian);
    return this;
  }

  /// Appends four bytes to the current contents of the builder with
  /// the four's complement binary representation of the specified value,
  /// which must fit in four bytes.
  BytesDataBuilder addUint32(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 32, false, endian);
    return this;
  }

  /// Appends eight bytes to the current contents of the builder with
  /// the eight's complement binary representation of the specified value,
  /// which must fit in eights bytes.
  BytesDataBuilder addUint64(int data,
      [Endianness endian = Endianness.BIG_ENDIAN]) {
    _add(data, 64, false, endian);
    return this;
  }
}


benchmark/ByteData_bench.dart
import 'dart:math';
import 'dart:typed_data';
import 'package:benchmark_harness/benchmark_harness.dart';

/// AUTHORS NOTE:
/// This runs the best, BUT you have to know the exact size of the data packet
/// you want to build before hand, by the time you calculate the size for this
/// ByteData instance, it may be more expensive in time to use a ByteData rather
/// then using the [BytesDataBuilder].
///
/// OUTPUT: (your times will vary)
/// Data(RunTime): 10.27395501035101 us.
/// Data(RunTime): 7.618147952051377 us.
/// Data(RunTime): 5.373859398342701 us.
/// Data(RunTime): 10.209811628975446 us.
/// Data(RunTime): 10.222544800302588 us.
/// Data(RunTime): 15.806278253722379 us.
///
/// NOTE:
/// The Dart VM can run in two modes: checked and production mode.
/// Checked mode is slower because the VM is checking types at runtime.
/// Before benchmarking make sure that your code runs without issue in checked mode.
/// If checked mode finds an issue, it will likely cause a performance problem
/// in production mode. After making sure your program is correct,
/// you should run your benchmark in production mode to get an accurate
/// measurement of real world performance.
///
/// Source: https://www.dartlang.org/articles/dart-vm/benchmarking

// Create a new benchmark by extending BenchmarkBase.
class DataBenchmark extends BenchmarkBase {
  DataBenchmark() : super("Data");

  int int8data;
  int int16data;
  int int32data;
  int int64data;
  int Uint8data;
  int Uint16data;
  int Uint32data;
  int Uint64data;
  List done;

  static void main() {
    new DataBenchmark().report();
  }

  // The benchmark code.
  void run() {
    ByteData data = new ByteData(60);
    int pos = 0;
    data.setInt8(pos, int8data); pos += 1;
    data.setInt16(pos, int16data); pos += 2;
    data.setInt32(pos, int32data); pos += 4;
    data.setInt64(pos, int64data); pos += 8;
    data.setUint8(pos, Uint8data); pos +=  1;
    data.setUint16(pos, Uint16data); pos += 2;
    data.setUint32(pos, Uint32data); pos += 4;
    data.setUint64(pos, Uint64data); pos += 8;
    data.setInt8(pos, int8data); pos += 1;
    data.setInt16(pos, int16data, Endianness.LITTLE_ENDIAN); pos += 2;
    data.setInt32(pos, int32data, Endianness.LITTLE_ENDIAN); pos += 4;
    data.setInt64(pos, int64data, Endianness.LITTLE_ENDIAN); pos += 8;
    data.setUint8(pos, Uint8data); pos += 1;
    data.setUint16(pos, Uint16data, Endianness.LITTLE_ENDIAN); pos += 2;
    data.setUint32(pos, Uint32data, Endianness.LITTLE_ENDIAN); pos += 4;
    data.setUint64(pos, Uint64data, Endianness.LITTLE_ENDIAN); pos += 8;
    done = data.buffer.asInt8List();
  }

  // Not measured: setup code executed before the benchmark runs.
  void setup() {
    Random rand = new Random();
    int8data = rand.nextInt(255) - 127;
    int16data = rand.nextInt(65535) - 32767;
    int32data = rand.nextInt(4294967295) - 2147483647;
    int rand64 = rand.nextInt(4294967295) * rand.nextInt(4294967295);
    int64data = rand64 - 9223372036854775807;
    Uint8data = rand.nextInt(255);
    Uint16data = rand.nextInt(65535);
    Uint32data = rand.nextInt(4294967295);
    Uint64data = rand.nextInt(4294967295) * rand.nextInt(4294967295);
  }

  // Not measured: teardown code executed after the benchmark runs.
  void teardown() {
  }
}

// Main function runs the benchmark.
main() {
  // Run DataBenchmark.
  DataBenchmark.main();
  DataBenchmark.main();
  DataBenchmark.main();
  DataBenchmark.main();
  DataBenchmark.main();
  DataBenchmark.main();
}

benchmark/DynamicByteData_bench.dart
import 'dart:math';
import 'dart:typed_data';
import 'package:benchmark_harness/benchmark_harness.dart';

/// AUTHORS NOTE:
/// Honestly quite surprised at how efficient the dart optimizer is.
/// This is the absolute worst way to build byte data, please don't ever do this.
///
/// OUTPUT: (your times will vary)
/// DynData(RunTime): 17.500721904778572 us.
/// DynData(RunTime): 26.935664166139176 us.
/// DynData(RunTime): 21.584985484097263 us.
/// DynData(RunTime): 18.657934753202166 us.
/// DynData(RunTime): 21.59570677349343 us.
/// DynData(RunTime): 21.90604497311033 us.
///
/// NOTE:
/// The Dart VM can run in two modes: checked and production mode.
/// Checked mode is slower because the VM is checking types at runtime.
/// Before benchmarking make sure that your code runs without issue in checked mode.
/// If checked mode finds an issue, it will likely cause a performance problem
/// in production mode. After making sure your program is correct,
/// you should run your benchmark in production mode to get an accurate
/// measurement of real world performance.
///
/// Source: https://www.dartlang.org/articles/dart-vm/benchmarking

// Create a new benchmark by extending BenchmarkBase.
class DynDataBenchmark extends BenchmarkBase {
  DynDataBenchmark() : super("DynData");

  int int8data;
  int int16data;
  int int32data;
  int int64data;
  int Uint8data;
  int Uint16data;
  int Uint32data;
  int Uint64data;
  List done;

  static void main() {
    new DynDataBenchmark().report();
  }

  // The benchmark code.
  void run() {
    List dat = [];
    ByteData data;
    /// overkill but you get my point
    data = new ByteData(1); data.setInt8(0, int8data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(2); data.setInt16(0, int16data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(4); data.setInt32(0, int32data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(8); data.setInt64(0, int64data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(1); data.setUint8(0, Uint8data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(2); data.setUint16(0, Uint16data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(4); data.setUint32(0, Uint32data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(8); data.setUint64(0, Uint64data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(1); data.setInt8(0, int8data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(2); data.setInt16(0, int16data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(4); data.setInt32(0, int32data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(8); data.setInt64(0, int64data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(1); data.setUint8(0, Uint8data); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(2); data.setUint16(0, Uint16data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(4); data.setUint32(0, Uint32data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    data = new ByteData(8); data.setUint64(0, Uint64data, Endianness.LITTLE_ENDIAN); dat.addAll(data.buffer.asInt8List());
    done = dat;
  }

  // Not measured: setup code executed before the benchmark runs.
  void setup() {
    Random rand = new Random();
    int8data = rand.nextInt(255) - 127;
    int16data = rand.nextInt(65535) - 32767;
    int32data = rand.nextInt(4294967295) - 2147483647;
    int rand64 = rand.nextInt(4294967295) * rand.nextInt(4294967295);
    int64data = rand64 - 9223372036854775807;
    Uint8data = rand.nextInt(255);
    Uint16data = rand.nextInt(65535);
    Uint32data = rand.nextInt(4294967295);
    Uint64data = rand.nextInt(4294967295) * rand.nextInt(4294967295);
  }

  // Not measured: teardown code executed after the benchmark runs.
  void teardown() {
  }
}

// Main function runs the benchmark.
main() {
  // Run DynDataBenchmark.
  DynDataBenchmark.main();
  DynDataBenchmark.main();
  DynDataBenchmark.main();
  DynDataBenchmark.main();
  DynDataBenchmark.main();
  DynDataBenchmark.main();
}

example/BytesDataBuilder_example.dart
// Copyright (c) 2016, Kurt Morris. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:math';
import 'package:bytes_data_builder/bytes_data_builder.dart';
import 'dart:typed_data';

main() {
  Random rand = new Random();
  int int8data = rand.nextInt(255) - 127;
  int int16data = rand.nextInt(65535) - 32767;
  int int32data = rand.nextInt(4294967295) - 2147483647;
  int rand64 = rand.nextInt(4294967295) * rand.nextInt(4294967295);
  int int64data = rand64 - 9223372036854775807;
  int Uint8data = rand.nextInt(255);
  int Uint16data = rand.nextInt(65535);
  int Uint32data = rand.nextInt(4294967295);
  int Uint64data = rand.nextInt(4294967295) * rand.nextInt(4294967295);

  BytesDataBuilder builder = new BytesDataBuilder();
  builder.addInt8(int8data);
  builder.addInt16(int16data);
  builder.addInt32(int32data);
  builder.addInt64(int64data);
  builder.addUint8(Uint8data);
  builder.addUint16(Uint16data);
  builder.addUint32(Uint32data);
  builder.addUint64(Uint64data);
  builder.addInt8(int8data);
  builder.addInt16(int16data, Endianness.LITTLE_ENDIAN);
  builder.addInt32(int32data, Endianness.LITTLE_ENDIAN);
  builder.addInt64(int64data, Endianness.LITTLE_ENDIAN);
  builder.addUint8(Uint8data);
  builder.addUint16(Uint16data, Endianness.LITTLE_ENDIAN);
  builder.addUint32(Uint32data, Endianness.LITTLE_ENDIAN);
  builder.addUint64(Uint64data, Endianness.LITTLE_ENDIAN);

  List<int> result = builder.takeBytes();
}


benchmark/BytesDataBuilder_bench.dart
import 'dart:math';
import 'dart:typed_data';
import 'package:bytes_data_builder/bytes_data_builder.dart';
import 'package:benchmark_harness/benchmark_harness.dart';

/// AUTHORS NOTE:
/// I'm a little surprised that this didn't bench better, but I think it may be
/// due to the fact that ByteData is written in Native code rather than pure dart.
/// I still think this is the best option for a growing BytesBuilder with other
/// integer options.
///
/// OUTPUT: (your times will vary)
/// DynData(RunTime): 20.101108576137975 us.
/// DynData(RunTime): 17.562346329469616 us.
/// DynData(RunTime): 17.279364119400405 us.
/// DynData(RunTime): 19.97642781517809 us.
/// DynData(RunTime): 15.116700931188777 us.
/// DynData(RunTime): 20.010405410813622 us.
///
/// NOTE:
/// The Dart VM can run in two modes: checked and production mode.
/// Checked mode is slower because the VM is checking types at runtime.
/// Before benchmarking make sure that your code runs without issue in checked mode.
/// If checked mode finds an issue, it will likely cause a performance problem
/// in production mode. After making sure your program is correct,
/// you should run your benchmark in production mode to get an accurate
/// measurement of real world performance.
///
/// Source: https://www.dartlang.org/articles/dart-vm/benchmarking

// Create a new benchmark by extending BenchmarkBase.
class BuilderBenchmark extends BenchmarkBase {
  BuilderBenchmark() : super("Builder");

  int int8data;
  int int16data;
  int int32data;
  int int64data;
  int Uint8data;
  int Uint16data;
  int Uint32data;
  int Uint64data;
  List done;

  static void main() {
    new BuilderBenchmark().report();
  }

  // The benchmark code.
  void run() {
    BytesDataBuilder builder = new BytesDataBuilder();
    builder.addInt8(int8data);
    builder.addInt16(int16data);
    builder.addInt32(int32data);
    builder.addInt64(int64data);
    builder.addUint8(Uint8data);
    builder.addUint16(Uint16data);
    builder.addUint32(Uint32data);
    builder.addUint64(Uint64data);
    builder.addInt8(int8data);
    builder.addInt16(int16data, Endianness.LITTLE_ENDIAN);
    builder.addInt32(int32data, Endianness.LITTLE_ENDIAN);
    builder.addInt64(int64data, Endianness.LITTLE_ENDIAN);
    builder.addUint8(Uint8data);
    builder.addUint16(Uint16data, Endianness.LITTLE_ENDIAN);
    builder.addUint32(Uint32data, Endianness.LITTLE_ENDIAN);
    builder.addUint64(Uint64data, Endianness.LITTLE_ENDIAN);
    // takeBytes is cheaper then toBytes.
    done = builder.takeBytes();
  }

  // Not measured: setup code executed before the benchmark runs.
  void setup() {
    Random rand = new Random();
    int8data = rand.nextInt(255) - 127;
    int16data = rand.nextInt(65535) - 32767;
    int32data = rand.nextInt(4294967295) - 2147483647;
    int rand64 = rand.nextInt(4294967295) * rand.nextInt(4294967295);
    int64data = rand64 - 9223372036854775807;
    Uint8data = rand.nextInt(255);
    Uint16data = rand.nextInt(65535);
    Uint32data = rand.nextInt(4294967295);
    Uint64data = rand.nextInt(4294967295) * rand.nextInt(4294967295);
  }

  // Not measured: teardown code executed after the benchmark runs.
  void teardown() {
  }
}

// Main function runs the benchmark.
main() {
  // Run BuilderBenchmark.
  BuilderBenchmark.main();
  BuilderBenchmark.main();
  BuilderBenchmark.main();
  BuilderBenchmark.main();
  BuilderBenchmark.main();
  BuilderBenchmark.main();
}